tests/proxy_test.go

```diff
@@ -219,6 +220,58 @@ func TestProxyHandle_SlowContext_GRPC(t *testing.T) {
 	}
 }
 
+func TestProxyHandle_ContextCancelled_GRPC(t *testing.T) {
+	defer goleak.VerifyNone(t)
```

Comment:
Without the bug fix in konnectivity-client/pkg/client/client.go:
```
$ go test ./... -run TestProxyHandle_ContextCancelled_GRPC
E0214 16:09:13.663053  847761 client.go:103] "stream read failure" err="rpc error: code = Canceled desc = context canceled"
E0214 16:09:13.663475  847761 server.go:390] "Stream read from frontend failure" err="rpc error: code = Canceled desc = context canceled"
E0214 16:09:13.664071  847761 server.go:751] "send to client stream failure" err="rpc error: code = Unavailable desc = transport is closing" serverID="846800f0-2f8e-4f29-b25a-50153d1f00cd" agentID="57a4766a-a392-4c4f-8146-c5a25b0c510f" connectionID=1
E0214 16:09:13.664223  847761 client.go:371] "could not read stream" err="rpc error: code = Unavailable desc = error reading from server: EOF"
E0214 16:09:13.664452  847761 server.go:515] "CLOSE_REQ to Backend failed" err="rpc error: code = Unavailable desc = transport is closing" serverID="846800f0-2f8e-4f29-b25a-50153d1f00cd"
E0214 16:09:13.664551  847761 server.go:669] "stream read failure" err="rpc error: code = Canceled desc = context canceled"
E0214 16:09:13.664551  847761 client.go:427] "close response failure" err="EOF"
E0214 16:09:13.665131  847761 server.go:705] "CLOSE_RSP to frontend failed" err="rpc error: code = Unavailable desc = transport is closing" serverID="846800f0-2f8e-4f29-b25a-50153d1f00cd" agentID="57a4766a-a392-4c4f-8146-c5a25b0c510f"
--- FAIL: TestProxyHandle_ContextCancelled_GRPC (2.45s)
    leaks.go:78: found unexpected goroutines:
        [Goroutine 19 in state chan receive, with k8s.io/klog/v2.(*loggingT).flushDaemon on top of the stack:
        goroutine 19 [chan receive]:
        k8s.io/klog/v2.(*loggingT).flushDaemon(0x0)
        	/home/andrewsy/go/pkg/mod/k8s.io/klog/v2@v2.4.0/klog.go:1169 +0x6a
        created by k8s.io/klog/v2.init.0
        	/home/andrewsy/go/pkg/mod/k8s.io/klog/v2@v2.4.0/klog.go:417 +0xfb
        
         Goroutine 31 in state chan receive, with sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client.(*conn).Read on top of the stack:
        goroutine 31 [chan receive]:
        sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client.(*conn).Read(0xc000485310, {0xc00025e000, 0x1000, 0x101000000000000})
        	/home/andrewsy/go/src/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client/conn.go:74 +0x5d
        net/http.(*persistConn).Read(0xc00015cea0, {0xc00025e000, 0x40717d, 0x60})
        	/usr/local/go/src/net/http/transport.go:1926 +0x4e
        bufio.(*Reader).fill(0xc0004de7e0)
        	/usr/local/go/src/bufio/bufio.go:101 +0x103
        bufio.(*Reader).Peek(0xc0004de7e0, 0x1)
        	/usr/local/go/src/bufio/bufio.go:139 +0x5d
        net/http.(*persistConn).readLoop(0xc00015cea0)
        	/usr/local/go/src/net/http/transport.go:2087 +0x1ac
        created by net/http.(*Transport).dialConn
        	/usr/local/go/src/net/http/transport.go:1747 +0x1e05
        ]
FAIL
FAIL	sigs.k8s.io/apiserver-network-proxy/tests	2.532s
FAIL
```

With the bug fix:
```
$ go test ./... -run TestProxyHandle_ContextCancelled_GRPC
E0214 16:16:42.596087  848745 client.go:112] "stream read failure" err="rpc error: code = Canceled desc = context canceled"
E0214 16:16:42.596265  848745 server.go:390] "Stream read from frontend failure" err="rpc error: code = Canceled desc = context canceled"
E0214 16:16:42.596361  848745 server.go:751] "send to client stream failure" err="rpc error: code = Canceled desc = context canceled" serverID="235e6686-f70f-4a1d-a6c6-dc46a811ce01" agentID="a947c418-4784-4b14-8fc9-c4cf0f47834b" connectionID=1
E0214 16:16:42.596473  848745 client.go:371] "could not read stream" err="rpc error: code = Unavailable desc = error reading from server: EOF"
E0214 16:16:42.596581  848745 server.go:515] "CLOSE_REQ to Backend failed" err="rpc error: code = Unavailable desc = transport is closing" serverID="235e6686-f70f-4a1d-a6c6-dc46a811ce01"
E0214 16:16:42.596624  848745 server.go:669] "stream read failure" err="rpc error: code = Canceled desc = context canceled"
E0214 16:16:42.596660  848745 client.go:427] "close response failure" err="EOF"
E0214 16:16:42.596694  848745 server.go:705] "CLOSE_RSP to frontend failed" err="rpc error: code = Unavailable desc = transport is closing" serverID="235e6686-f70f-4a1d-a6c6-dc46a811ce01" agentID="a947c418-4784-4b14-8fc9-c4cf0f47834b"
--- FAIL: TestProxyHandle_ContextCancelled_GRPC (2.45s)
    leaks.go:78: found unexpected goroutines:
        [Goroutine 6 in state chan receive, with k8s.io/klog/v2.(*loggingT).flushDaemon on top of the stack:
        goroutine 6 [chan receive]:
        k8s.io/klog/v2.(*loggingT).flushDaemon(0x0)
        	/home/andrewsy/go/pkg/mod/k8s.io/klog/v2@v2.4.0/klog.go:1169 +0x6a
        created by k8s.io/klog/v2.init.0
        	/home/andrewsy/go/pkg/mod/k8s.io/klog/v2@v2.4.0/klog.go:417 +0xfb
        ]
FAIL
FAIL	sigs.k8s.io/apiserver-network-proxy/tests	2.528s
FAIL
```

Note that the test always fails with 1 goroutine leak coming from klog which is expected.

Reply:
What if we as soon as this test function starts, or if we can initialize klog early first, then we use [IgnoreCurrent](https://pkg.go.dev/go.uber.org/goleak#IgnoreCurrent) to make it ignore klog and keep the test?

It seems super useful this pkg, I'd love to keep it in all tests. 

(btw, I wasn't aware of this module, seems quite cool. Thanks!!)

Reply:
Done

Reply:
And just verifying the test works:

Without defer close(conn.readCh)
```
$ go test ./... -run TestProxyHandle_ContextCancelled_GRPC
E0216 09:32:28.416716 1191438 client.go:103] "stream read failure" err="rpc error: code = Canceled desc = context canceled"
E0216 09:32:28.417368 1191438 server.go:751] "send to client stream failure" err="rpc error: code = Unavailable desc = transport is closing" serverID="acbcdbed-f7a1-40b5-a02c-c65186c172a0" agentID="b441b2f6-c9c1-43a9-9f33-f30d1fb1306e" connectionID=1
E0216 09:32:28.417398 1191438 server.go:390] "Stream read from frontend failure" err="rpc error: code = Canceled desc = context canceled"
E0216 09:32:28.417907 1191438 client.go:371] "could not read stream" err="rpc error: code = Unavailable desc = error reading from server: read tcp [::1]:39060->[::1]:40799: read: connection reset by peer"
E0216 09:32:28.419031 1191438 client.go:427] "close response failure" err="EOF"
E0216 09:32:28.419143 1191438 server.go:515] "CLOSE_REQ to Backend failed" err="rpc error: code = Unavailable desc = transport is closing" serverID="acbcdbed-f7a1-40b5-a02c-c65186c172a0"
E0216 09:32:28.419955 1191438 server.go:669] "stream read failure" err="rpc error: code = Canceled desc = context canceled"
E0216 09:32:28.420246 1191438 server.go:705] "CLOSE_RSP to frontend failed" err="rpc error: code = Unavailable desc = transport is closing" serverID="acbcdbed-f7a1-40b5-a02c-c65186c172a0" agentID="b441b2f6-c9c1-43a9-9f33-f30d1fb1306e"
--- FAIL: TestProxyHandle_ContextCancelled_GRPC (2.46s)
    leaks.go:78: found unexpected goroutines:
        [Goroutine 100 in state chan receive, with sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client.(*conn).Read on top of the stack:
        goroutine 100 [chan receive]:
        sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client.(*conn).Read(0xc00003d450, {0xc0001a7000, 0x1000, 0x101000000000000})
        	/home/andrewsy/go/src/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client/conn.go:74 +0x5d
        net/http.(*persistConn).Read(0xc0004207e0, {0xc0001a7000, 0x40717d, 0x60})
        	/usr/local/go/src/net/http/transport.go:1926 +0x4e
        bufio.(*Reader).fill(0xc000099e00)
        	/usr/local/go/src/bufio/bufio.go:101 +0x103
        bufio.(*Reader).Peek(0xc000099e00, 0x1)
        	/usr/local/go/src/bufio/bufio.go:139 +0x5d
        net/http.(*persistConn).readLoop(0xc0004207e0)
        	/usr/local/go/src/net/http/transport.go:2087 +0x1ac
        created by net/http.(*Transport).dialConn
        	/usr/local/go/src/net/http/transport.go:1747 +0x1e05
        ]
FAIL
FAIL	sigs.k8s.io/apiserver-network-proxy/tests	2.549s
FAIL
```

With defer close(conn.readCh):
```
$ go test ./... -run TestProxyHandle_ContextCancelled_GRPC -count 1
ok  	sigs.k8s.io/apiserver-network-proxy/tests	2.222s
```

Reply:
Awesome, thanks! :)